src/dynarmic/interface/tlb.h

```diff
@@ -0,0 +1,122 @@
+/* This file is part of the dynarmic project.
+ * Copyright (c) 2018 MerryMage
+ * SPDX-License-Identifier: 0BSD
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <cstring>
+#include <vector>
+
+namespace Dynarmic {
+
+using VAddr = std::uint64_t;
+
+enum MemoryPermission {
+    MemoryPermissionRead = 1 << 0,
+    MemoryPermissionWrite = 1 << 1,
+    MemoryPermissionExecute = 1 << 2
+};
+
+struct TLBEntry {
+    VAddr read_addr;
+    VAddr write_addr;
+    VAddr execute_addr;
+
+    std::uint8_t *host_base;
+
+    explicit TLBEntry()
+        : read_addr(0)
+        , write_addr(0)
+        , execute_addr(0)
+        , host_base(nullptr) {
+    }
+};
+
+template <std::uint32_t TLB_BIT_COUNT>
+struct TLB {
+public:
+    static constexpr std::uint32_t TLB_ENTRY_COUNT = 1 << TLB_BIT_COUNT;
+    static constexpr std::uint32_t TLB_BIT_MASK = (1 << TLB_BIT_COUNT) - 1;
+
+    TLBEntry entries[TLB_ENTRY_COUNT];
+
+    std::size_t page_bits;
+    std::size_t page_mask;
+
+    explicit TLB(std::size_t page_bits)
+        : page_bits(page_bits) {
```

Comment:
The assignment can be moved up into the initializer list
```suggestion
        : page_bits(page_bits), page_mask((size_t{1} << page_bits) - 1) {
```